python/ql/lib/semmle/python/frameworks/Gradio.qll

```diff
@@ -0,0 +1,157 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `gradio` PyPI package.
+ * See https://pypi.org/project/gradio/.
+ */
+
+import python
+import semmle.python.dataflow.new.RemoteFlowSources
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.ApiGraphs
+
+/**
+ * Provides models for the `gradio` PyPI package.
+ * See https://pypi.org/project/gradio/.
+ */
+module Gradio {
+  /**
+   * The event handlers in Gradio, which take untrusted data.
+   */
+  class GradioInput extends API::CallNode {
+    GradioInput() {
+      this =
+        API::moduleImport("gradio")
+            .getMember([
+                "Button", "Textbox", "UploadButton", "Slider", "JSON", "HTML", "Markdown", "File",
+                "AnnotatedImage", "Audio", "BarPlot", "Chatbot", "Checkbox", "CheckboxGroup",
+                "ClearButton", "Code", "ColorPicker", "Dataframe", "Dataset", "DownloadButton",
+                "Dropdown", "DuplicateButton", "FileExplorer", "Gallery", "HighlightedText",
+                "Image", "ImageEditor", "Label", "LinePlot", "LoginButton", "LogoutButton",
+                "Model3D", "Number", "ParamViewer", "Plot", "Radio", "ScatterPlot", "SimpleImage",
+                "State", "Video"
+              ])
+            .getReturn()
+            .getMember([
+                "change", "input", "click", "submit", "edit", "clear", "play", "pause", "stop",
+                "end", "start_recording", "pause_recording", "stop_recording", "focus", "blur",
+                "upload", "release", "select", "stream", "like", "load", "like", "key_up",
+              ])
+            .getACall()
+    }
+  }
+
+  /**
+   * The high-level gradio.Interface and gradio.ChatInterface classes, which take untrusted data.
+   */
+  class GradioInterface extends API::CallNode {
+    GradioInterface() {
+      this = API::moduleImport("gradio").getMember(["Interface", "ChatInterface"]).getACall()
+    }
+  }
+
+  /**
+   * The `inputs` parameters in Gradio event handlers, that are lists and are sources of untrusted data.
+   * This model allows tracking each element list back to source, f.ex. `gr.Textbox(...)`.
+   */
+  class GradioInputList extends RemoteFlowSource::Range {
+    GradioInputList() {
+      exists(API::CallNode call |
+        (
+          call instanceof GradioInput
+          or
+          call instanceof GradioInterface
+        ) and
```

Comment:
Makes sense, I put the models from GradioInterface into GradioInput in https://github.com/github/codeql/pull/16135/commits/8bb41937047fc7a708d7c331d0881d2ce63dadb5, since they are used together all the time